kolibri/plugins/management/assets/src/state/addCoachRoleAction.js

```diff
@@ -0,0 +1,57 @@
+/* eslint-disable prefer-arrow-callback */
+import * as coreApp from 'kolibri';
+import * as constants from 'kolibri.coreVue.vuex.constants';
+
+const { RoleResource } = coreApp.resources;
+const { COACH } = constants.UserKinds;
+
+export function dispatchError(store, err) {
+  return store.dispatch('CORE_SET_ERROR', JSON.stringify(err.entity || err.message));
+}
+
+export function dispatchRoleChange(store, payload) {
+  store.dispatch('UPDATE_LEARNER_ROLE_FOR_CLASS', payload);
+}
+
+/**
+ * Adds a Role to a User in the context of a Collection
+ * Generalizes simililar Role factories that use Facility as Collection concept
+ * This is not an idempotent action. If you try to do the same thing twice, you get an Error.
+ * @param {Object} payload
+ * @param {string} payload.userId
+ * @param {string} payload.collectionId
+ * @param {string} payload.newRole - maps to `kind`
+ * @returns {Promise}
+ */
+function addRoleToUserInCollection(payload) {
+  return RoleResource.createModel({
+    user: payload.userId,
+    collection: payload.collectionId,
+    kind: payload.newRole,
+  }).save();
+}
+
+/**
+ * Vuex action that adds the 'coach' role to a Learner in a Class and updates store accordingly
+ * @param {Object} payload
+ * @param {string} payload.userId
+ * @param {string} payload.classId
+ * @returns {Promise}
+ */
+export default function addCoachRoleAction(store, payload) {
+  const { classId, userId } = payload;
+  const newRole = COACH;
+  return (
+    addRoleToUserInCollection({
+      collectionId: classId,
+      newRole,
+      userId,
+    })
+    .then(function onSuccess() {
+      dispatchRoleChange(store, { newRole, userId });
+    })
+    .catch(function onFailure(err) {
+      dispatchError(store, err);
+    })
```

Comment:
Currently, the UI waits until a server response to handle either success or failure. This leaves open the possibility that the UI might feel unresponsive on a slow connection, or even produce an error if the user hits the toggle switch twice. (You can demo this by turning on network throttling in chrome tools.)

Simple solution: disable the toggle switch while the response is pending

More complex: optimistically update vuex and the UI with the intended new value immediately on-click. A success becomes a no-op, and an error would probably roll back to the previous value.

Another thing to watch out for is if the user manages to navigate away from the page before the response comes back. In other actions, we've tried to leverage the `ConditionalPromise` to handle this case. The idea is that success or failure handlers are only called based on a certain condition, typically that the user is on the same same page. (See `coreActions.samePageCheckGenerator`)